# Exploring the Dataset:

In [126]:
# import statements:

import pandas as pd

# Exploring the Dataset:

In [127]:
# reading the dataset with pandas and turning it into a pandas DF:

df = pd.read_csv('../data/googleplaystore.csv')

Having a general look at the dataset:


In [128]:
# looking at the numer of rows and columns in the dataset:

df.shape

(10841, 13)

In [129]:
# looking at the dataframe'c columns and the dtypes:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


we have some null values in the dataframe!

In [130]:
# looking at the first 5 rows of the dataframe to see how the columns are filled:

df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

### Cleaning up the dataframe:

#### Duplicated Rows:

In [131]:
# checking the number of duplicated rows:

df.duplicated().sum()

483

we have 483 duplicated rows, so we need to drop them entirely.

In [132]:
#dropping the duplicated rows and storing the resulting entities in a new dataframe:

clean_df = df.drop_duplicates()

In [133]:
# making sure that we got rid of the duplicated rows:

clean_df.duplicated().sum()

0

let us check for duplicates in the app names and see if they are different entities:

In [134]:
clean_df['App'].duplicated().sum()

698

We have 698 duplicated Apps, let's check them out!

In [135]:
clean_df[clean_df['App'].duplicated() == True]['App'].value_counts()

App
ROBLOX                            8
8 Ball Pool                       6
Bubble Shooter                    5
Zombie Catchers                   5
Helix Jump                        5
                                 ..
Google Docs                       1
Microsoft Excel                   1
Smart Launcher 5                  1
Calculator                        1
Slickdeals: Coupons & Shopping    1
Name: count, Length: 523, dtype: int64

In [136]:
clean_df[clean_df['App']== 'ROBLOX']

App Category  Rating  Reviews Size      Installs  Type Price  \
1653  ROBLOX     GAME     4.5  4447388  67M  100,000,000+  Free     0   
1701  ROBLOX     GAME     4.5  4447346  67M  100,000,000+  Free     0   
1748  ROBLOX     GAME     4.5  4448791  67M  100,000,000+  Free     0   
1841  ROBLOX     GAME     4.5  4449882  67M  100,000,000+  Free     0   
1870  ROBLOX     GAME     4.5  4449910  67M  100,000,000+  Free     0   
2016  ROBLOX   FAMILY     4.5  4449910  67M  100,000,000+  Free     0   
2088  ROBLOX   FAMILY     4.5  4450855  67M  100,000,000+  Free     0   
2206  ROBLOX   FAMILY     4.5  4450890  67M  100,000,000+  Free     0   
4527  ROBLOX   FAMILY     4.5  4443407  67M  100,000,000+  Free     0   

     Content Rating                        Genres   Last Updated  \
1653   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1701   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1748   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1841   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
1870   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
2016   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
2088   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
2206   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   
4527   Everyone 10+  Adventure;Action & Adventure  July 31, 2018   

       Current Ver Android Ver  
1653  2.347.225742  4.1 and up  
1701  2.347.225742  4.1 and up  
1748  2.347.225742  4.1 and up  
1841  2.347.225742  4.1 and up  
1870  2.347.225742  4.1 and up  
2016  2.347.225742  4.1 and up  
2088  2.347.225742  4.1 and up  
2206  2.347.225742  4.1 and up  
4527  2.347.225742  4.1 and up

In [137]:
clean_df[clean_df['App']== '8 Ball Pool']

App Category  Rating   Reviews Size      Installs  Type Price  \
1675  8 Ball Pool     GAME     4.5  14198297  52M  100,000,000+  Free     0   
1703  8 Ball Pool     GAME     4.5  14198602  52M  100,000,000+  Free     0   
1755  8 Ball Pool     GAME     4.5  14200344  52M  100,000,000+  Free     0   
1844  8 Ball Pool     GAME     4.5  14200550  52M  100,000,000+  Free     0   
1871  8 Ball Pool     GAME     4.5  14201891  52M  100,000,000+  Free     0   
1970  8 Ball Pool     GAME     4.5  14201604  52M  100,000,000+  Free     0   
3953  8 Ball Pool   SPORTS     4.5  14184910  52M  100,000,000+  Free     0   

     Content Rating  Genres   Last Updated Current Ver   Android Ver  
1675       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1703       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1755       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1844       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1871       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
1970       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up  
3953       Everyone  Sports  July 31, 2018       4.0.0  4.0.3 and up

In [138]:
clean_df[clean_df['App']== 'Google Docs']

App      Category  Rating Reviews                Size  \
3477  Google Docs  PRODUCTIVITY     4.3  815981  Varies with device   
3528  Google Docs  PRODUCTIVITY     4.3  815974  Varies with device   

          Installs  Type Price Content Rating        Genres    Last Updated  \
3477  100,000,000+  Free     0       Everyone  Productivity  August 6, 2018   
3528  100,000,000+  Free     0       Everyone  Productivity  August 6, 2018   

             Current Ver         Android Ver  
3477  Varies with device  Varies with device  
3528  Varies with device  Varies with device

We can see that the apps differ mostly by the number of reviews and categories. I will drop the duplicates according to the number of reviews. 

We will be keeping the entry with the highest number of reviews!

To do so we have to go through a number of steps:

In [139]:
# sorting the df by app name and number of reviews:

clean_df = clean_df.sort_values(by=['App', 'Reviews'], ascending=[True, False])

In [140]:
# dropping duplicates on the App name level and just keeping the first entry which will be the one with the highest number of reviews:

clean_df = clean_df.drop_duplicates(subset=['App'], keep='first')

In [141]:
# making sure this worked:

clean_df['App'].duplicated().sum()

0

now that we are done with duplicates, let us see what we can do with the null values:

### Null Values:

In [142]:
# getting the number of null values in the dataframe:

clean_df.isna().sum()

App                  0
Category             0
Rating            1463
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

let us look first at the lesser null values:

In [143]:
# getting the rows with null values in the Type column:

clean_df[clean_df['Type'].isna()==True]

App Category  Rating Reviews                Size  \
9148  Command & Conquer: Rivals   FAMILY     NaN       0  Varies with device   

     Installs Type Price Content Rating    Genres   Last Updated  \
9148        0  NaN     0   Everyone 10+  Strategy  June 28, 2018   

             Current Ver         Android Ver  
9148  Varies with device  Varies with device

we can fill this by external research!

With research, we found out that this game is free!

In [144]:
# replacing the null value with the app type!

clean_df.loc[clean_df['App'] == 'Command & Conquer: Rivals', 'Type'] = 'Free'


In [145]:
# making sure the previous step worked

clean_df['Type'].isna().sum()

0

In [146]:
# getting the rows with null values in the Content Rating column:

clean_df[clean_df['Content Rating'].isna()==True]

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

In [147]:
# replacing the null value:

clean_df.loc[clean_df['App'] == 'Life Made WI-Fi Touchscreen Photo Frame', 'Content Rating'] = 'Everyone'

In [148]:
# checking if it worked:

clean_df['Content Rating'].isna().sum()

0

In [149]:
# getting the rows with null values in the Curr Ver. column:

clean_df[clean_df['Current Ver'].isna()==True]

App             Category  Rating Reviews  \
6803                         BT Master               FAMILY     NaN       0   
7407                   Calculate My IQ               FAMILY     NaN      44   
7333                       Dots puzzle               FAMILY     4.0     179   
10342                   La Fe de Jesus  BOOKS_AND_REFERENCE     NaN       8   
15     Learn To Draw Kawaii Characters       ART_AND_DESIGN     3.2      55   
1553              Market Update Helper   LIBRARIES_AND_DEMO     4.1   20145   
7730                            UFO-CQ                TOOLS     NaN       1   
6322            Virtual DJ Sound Mixer                TOOLS     4.2    4010   

       Size    Installs  Type  Price Content Rating             Genres  \
6803   222k        100+  Free      0       Everyone          Education   
7407   7.2M     10,000+  Free      0       Everyone      Entertainment   
7333    14M     50,000+  Paid  $0.99       Everyone             Puzzle   
10342  658k      1,000+  Free      0       Everyone  Books & Reference   
15     2.7M      5,000+  Free      0       Everyone       Art & Design   
1553    11k  1,000,000+  Free      0       Everyone   Libraries & Demo   
7730   237k         10+  Paid  $0.99       Everyone              Tools   
6322   8.7M    500,000+  Free      0       Everyone              Tools   

            Last Updated Current Ver Android Ver  
6803    November 6, 2016         NaN  1.6 and up  
7407       April 3, 2017         NaN  2.3 and up  
7333      April 18, 2018         NaN  4.0 and up  
10342   January 31, 2017         NaN  3.0 and up  
15          June 6, 2018         NaN  4.2 and up  
1553   February 12, 2013         NaN  1.5 and up  
7730        July 4, 2016         NaN  2.0 and up  
6322        May 10, 2017         NaN  4.0 and up

This will be hard to get and fill so we will drop the null values!

Also for Android Ver!

In [150]:
# dropping the null values in the two specified columns:

clean_df = clean_df.dropna(subset=['Current Ver', 'Android Ver'])

In [151]:
# checking if it worked:

clean_df.isna().sum()

App                  0
Category             0
Rating            1459
Reviews              0
Size                 0
Installs             0
Type                 0
Price                0
Content Rating       0
Genres               0
Last Updated         0
Current Ver          0
Android Ver          0
dtype: int64

Now to see what to do with the rating column as the null values are 15% of the dataframe!

In [152]:
clean_df[clean_df['Rating'].isna()==True]

App       Category  \
8884                "i DT" Fútbol. Todos Somos Técnicos.         SPORTS   
2483                                       14thStreetVet        MEDICAL   
7382                           1st Fed CI Mobile Banking        FINANCE   
6518                         2017 BN SM Sales Conference       BUSINESS   
9660                                      23rd QM BDE EO       BUSINESS   
...                                                  ...            ...   
321    【Ranobbe complete free】 Novelba - Free app tha...         COMICS   
9466                     パーリーゲイツ公式通販｜EJ STYLE（イージェイスタイル）       SHOPPING   
5513                                         中国語 AQリスニング         FAMILY   
10109                                            哈哈姆特不EY  COMMUNICATION   
6334                             뽕티비 - 개인방송, 인터넷방송, BJ방송  VIDEO_PLAYERS   

       Rating Reviews  Size  Installs  Type Price Content Rating  \
8884      NaN      27  3.6M      500+  Free     0       Everyone   
2483      NaN       0   29M        5+  Free     0       Everyone   
7382      NaN       0   31M      500+  Free     0       Everyone   
6518      NaN       2   10M      100+  Free     0           Teen   
9660      NaN       0  6.3M       10+  Free     0       Everyone   
...       ...     ...   ...       ...   ...   ...            ...   
321       NaN    1330   22M   50,000+  Free     0       Everyone   
9466      NaN       1  9.3M      100+  Free     0       Everyone   
5513      NaN      21   17M    5,000+  Free     0       Everyone   
10109     NaN     239   18M   10,000+  Free     0       Everyone   
6334      NaN     414   59M  100,000+  Free     0     Mature 17+   

                        Genres        Last Updated  Current Ver   Android Ver  
8884                    Sports     October 7, 2017         0.22    4.1 and up  
2483                   Medical       July 16, 2018  300000.1.11  4.0.3 and up  
7382                   Finance  September 29, 2016   6.0.0.1775    4.1 and up  
6518                  Business  September 28, 2017    v2.7.11.6    4.1 and up  
9660                  Business  September 27, 2017        1.0.0    4.1 and up  
...                        ...                 ...          ...           ...  
321                     Comics        July 3, 2018        6.1.1    4.2 and up  
9466                  Shopping   November 16, 2017          3.0  4.0.3 and up  
5513                 Education       June 22, 2016        2.4.0    4.0 and up  
10109            Communication       July 31, 2018        1.3.4    4.4 and up  
6334   Video Players & Editors       July 18, 2018        4.0.7  4.0.3 and up  

[1459 rows x 13 columns]

Sadly, we have to also drop the nulls in this column.

In [153]:
# dropping the nulls in the Rating column:

clean_df = clean_df.dropna(subset=['Rating'])

In [154]:
# checking if we have any null values left:

clean_df.isna().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              0
Price             0
Content Rating    0
Genres            0
Last Updated      0
Current Ver       0
Android Ver       0
dtype: int64

No more null values!

Now we can move to the next step which is fixing the columns and their data types!

### Fixing Up Columns + Dtypes:

In [155]:
# having a general look at the data again:

clean_df.head()

App            Category  Rating Reviews  Size  \
8532  +Download 4 Instagram Twitter              SOCIAL     4.5   40467   22M   
324      - Free Comics - Comic Apps              COMICS     3.5     115  9.1M   
4541                             .R               TOOLS     4.5     259  203k   
4636                         /u/app       COMMUNICATION     4.7     573   53M   
5940                         058.ba  NEWS_AND_MAGAZINES     4.4      27   14M   

        Installs  Type Price Content Rating            Genres  \
8532  1,000,000+  Free     0       Everyone            Social   
324      10,000+  Free     0     Mature 17+            Comics   
4541     10,000+  Free     0       Everyone             Tools   
4636     10,000+  Free     0     Mature 17+     Communication   
5940        100+  Free     0       Everyone  News & Magazines   

            Last Updated Current Ver Android Ver  
8532      August 2, 2018        5.03  4.1 and up  
324        July 13, 2018      5.0.12  5.0 and up  
4541  September 16, 2014      1.1.06  1.5 and up  
4636        July 3, 2018       4.2.4  4.1 and up  
5940        July 6, 2018         1.0  4.2 and up

#### Category Column:

In [156]:
# counting the number of unique categories:

clean_df['Category'].nunique()

33

In [157]:
# getting the names of the categories:

clean_df['Category'].value_counts()

Category
FAMILY                 1650
GAME                    896
TOOLS                   718
FINANCE                 302
LIFESTYLE               301
PRODUCTIVITY            301
PERSONALIZATION         296
MEDICAL                 290
PHOTOGRAPHY             263
BUSINESS                263
SPORTS                  260
COMMUNICATION           256
HEALTH_AND_FITNESS      244
NEWS_AND_MAGAZINES      204
SOCIAL                  203
TRAVEL_AND_LOCAL        187
SHOPPING                180
BOOKS_AND_REFERENCE     169
VIDEO_PLAYERS           149
DATING                  134
MAPS_AND_NAVIGATION     118
EDUCATION               107
FOOD_AND_DRINK           94
ENTERTAINMENT            87
AUTO_AND_VEHICLES        73
WEATHER                  72
LIBRARIES_AND_DEMO       63
HOUSE_AND_HOME           61
ART_AND_DESIGN           58
COMICS                   54
PARENTING                50
EVENTS                   45
BEAUTY                   42
Name: count, dtype: int64

we can see in some categories that the space is replaced with a '_', let us fix that:

In [158]:
# replacing '_' with ' ':

clean_df['Category'] = clean_df['Category'].str.replace('_', ' ')

In [159]:
# checking if it worked:

clean_df['Category'].value_counts()

Category
FAMILY                 1650
GAME                    896
TOOLS                   718
FINANCE                 302
LIFESTYLE               301
PRODUCTIVITY            301
PERSONALIZATION         296
MEDICAL                 290
PHOTOGRAPHY             263
BUSINESS                263
SPORTS                  260
COMMUNICATION           256
HEALTH AND FITNESS      244
NEWS AND MAGAZINES      204
SOCIAL                  203
TRAVEL AND LOCAL        187
SHOPPING                180
BOOKS AND REFERENCE     169
VIDEO PLAYERS           149
DATING                  134
MAPS AND NAVIGATION     118
EDUCATION               107
FOOD AND DRINK           94
ENTERTAINMENT            87
AUTO AND VEHICLES        73
WEATHER                  72
LIBRARIES AND DEMO       63
HOUSE AND HOME           61
ART AND DESIGN           58
COMICS                   54
PARENTING                50
EVENTS                   45
BEAUTY                   42
Name: count, dtype: int64

#### Reviews Column:

we just have to change the data type of the reviews column!

In [160]:
# changing the data type of the reviews column into an integer:

clean_df['Reviews'] = clean_df['Reviews'].astype(int)

#### Price Column:

We fix it up first removing the $ sign ang then tunring the price into a numeric

In [161]:
# removing the $ sign

clean_df['Price'] = clean_df['Price'].str.replace('$', '')

In [162]:
# changing the price column dtype to float:

clean_df['Price'] = clean_df['Price'].astype(float)